sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -280,61 +295,60 @@ class GpuDynamicPartitionDataSingleWriter(
     taskAttemptContext: TaskAttemptContext,
     committer: FileCommitProtocol)
   extends GpuFileFormatDataWriter(description, taskAttemptContext, committer) {
+  /** Wrapper class to index a unique concurrent output writer. */
+  protected case class WriterIndex(
+      var partitionPath: Option[String],
+      var bucketId: Option[Int]) {
+
+    override def hashCode(): Int = ScalaMurmur3Hash.productHash(this)
+
+    override def equals(obj: Any): Boolean = {
+      if (this.eq(obj.asInstanceOf[AnyRef])) {
+        true
+      } else if (obj.isInstanceOf[WriterIndex]) {
+        obj.hashCode() == this.hashCode()
+      } else {
+        false
+      }
+    }
+  }
 
-  /** Wrapper class for status of a unique single output writer. */
-  protected class WriterStatus(
-      // output writer
-      var outputWriter: ColumnarOutputWriter,
-
-      /** Number of records in current file. */
-      var recordsInFile: Long = 0,
-
-      /**
-       * File counter for writing current partition or bucket. For same partition or bucket,
-       * we may have more than one file, due to number of records limit per file.
-       */
-      var fileCounter: Int = 0
-  )
-
-  /** Wrapper class for status and caches of a unique concurrent output writer.
-   * Used by `GpuDynamicPartitionDataConcurrentWriter`
+  /**
+   * A case class to hold the batch, the optional partition path and the optional bucket
+   * ID for a split group. All the rows in the batch belong to the group defined by the
+   * partition path and the bucket ID.
    */
-  class WriterStatusWithCaches(
-      // writer status
-      var writerStatus: WriterStatus,
-
-      // caches for this partition or writer
-      val tableCaches: ListBuffer[SpillableColumnarBatch] = ListBuffer(),
-
-      // current device bytes for the above caches
-      var deviceBytes: Long = 0
-  )
+  private case class SplitPack(var split: SpillableColumnarBatch, path: Option[String],
```

Comment:
Why is this a case class and split is a var? Those two should be really rare. Can we please make this a regular class instead?

Reply:
Changed split to a val.